types/crossfilter2.d.ts

```diff
@@ -0,0 +1,103 @@
+declare namespace CrossFilter2 {
+
+    export interface Selector<T> {
+        (value: T): any;
+    }
+
+    export interface CrossFilterStatic {
+        <T>(data: T[]): CrossFilter<T>;
+        size(): number;
+        version: string;
+        permute<T>(arry: T[], index: number[], deep: number): T[];
+        heap: {
+            <T>(array: T[], lo: number, hi: number): T[];
+            by<T>(value: Selector<T>): Heap<T>;
+        }
+        heapselect: {
+            <T>(array: T[], lo: number, hi: number, k: number): T[];
+            by<T>(value: Selector<T>): HeapSelect<T>;
+        }
+        bisect: {
+            <T>(array: T[], value: T, lo: number, hi: number): number;
+            by<T>(value: Selector<T>): Bisector<T>;
+        }
+        insertionsort: {
+            <T>(array: T[], lo: number, hi: number): T[];
+            by<T>(vaule: Selector<T>): Sort<T>;
+        }
+        quicksort: {
+            <T>(array: T[], lo: number, hi: number): T[];
+            by<T>(value: Selector<T>): Sort<T>;
+        }
+    }
+
+    export interface Bisection<T> {
+        (array: T[], value: T, lo: number, hi: number): number;
+    }
+
+    export interface Bisector<T> extends Bisection<T> {
+        left: Bisection<T>
+        right: Bisection<T>
+    }
+
+    export interface Heap<T> {
+        (array: T[], lo: number, hi: number): T[];
+        sort(array: T[], lo: number, hi: number): T[];
+    }
+
+    export interface HeapSelect<T> {
+        (array: T[], lo: number, hi: number, k: number): T[];
+    }
+
+    export interface Sort<T> {
+        (array: T[], lo: number, hi: number): T[];
+    }
+
+    export interface CrossFilter<T> {
+        add(records: T[]): CrossFilter<T>;
+        remove(predicate?: Function): void;
+        dimension<T>(value: Function, isArray?: boolean): Dimension<T>;
+        groupAll(): any; // return vale?
+        size(): number;
+        all(): Array<T>;
+        allFiltered(): Array<T>;
+        onChange(callback: Function): Function;
+        isElementFiltered(index: number, ignoreDimensions?: number): boolean;
+
+    }
+
+    export interface Dimension<T> {
+        filter(range: any): any; // returns Array?
```

Comment:
I think all the `filter*` functions return `Dimension<T>`. The argument for `filter` would have the type that is the union of the types of the arguments for all the other filters.